pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -1447,76 +1348,182 @@ func (n *NetworkPolicyController) getInternalGroupWorkloads(group *antreatypes.G
+			return nil
+		}
+		newInternalNetworkPolicy, newAppliedToGroups, newAddressGroups = n.processNetworkPolicy(knp)
+	}
+
+	newNodeNames := sets.NewString()
 	// Calculate the set of Node names based on the span of the
 	// AppliedToGroups referenced by this NetworkPolicy.
-	for _, appliedToGroupName := range internalNP.AppliedToGroups {
+	for appliedToGroupName := range newAppliedToGroups {
 		appGroupObj, found, _ := n.appliedToGroupStore.Get(appliedToGroupName)
 		if !found {
 			continue
 		}
 		appGroup := appGroupObj.(*antreatypes.AppliedToGroup)
-		utilsets.MergeString(nodeNames, appGroup.SpanMeta.NodeNames)
-	}
-	updatedNetworkPolicy := &antreatypes.NetworkPolicy{
-		UID:                   internalNP.UID,
-		Name:                  internalNP.Name,
-		SourceRef:             internalNP.SourceRef,
-		Rules:                 internalNP.Rules,
-		AppliedToGroups:       internalNP.AppliedToGroups,
-		Priority:              internalNP.Priority,
-		TierPriority:          internalNP.TierPriority,
-		AppliedToPerRule:      internalNP.AppliedToPerRule,
-		PerNamespaceSelectors: internalNP.PerNamespaceSelectors,
-		SpanMeta:              antreatypes.SpanMeta{NodeNames: nodeNames},
-		Generation:            internalNP.Generation,
-		RealizationError:      internalNP.RealizationError,
-	}
-	klog.V(4).Infof("Updating internal NetworkPolicy %s with %d Nodes", key, nodeNames.Len())
-	n.internalNetworkPolicyStore.Update(updatedNetworkPolicy)
-	// Internal NetworkPolicy update is complete. Safe to unlock the
-	// critical section.
-	n.internalNetworkPolicyMutex.Unlock()
-	if nodeNames.Equal(oldNodeNames) {
-		// Node span for internal NetworkPolicy was not modified. No need to enqueue
-		// AddressGroups.
-		klog.V(4).Infof("Internal NetworkPolicy %s Node span remains unchanged. No need to enqueue AddressGroups.", key)
-		return nil
+		utilsets.MergeString(newNodeNames, appGroup.SpanMeta.NodeNames)
 	}
-	// Enqueue addressGroup keys to update their Node span.
-	for _, rule := range internalNP.Rules {
-		for _, addrGroupName := range rule.From.AddressGroups {
-			n.enqueueAddressGroup(addrGroupName)
+	newInternalNetworkPolicy.NodeNames = newNodeNames
+
+	var oldInternalNetworkPolicy *antreatypes.NetworkPolicy
+	oldInternalNetworkPolicyObj, oldInternalPolicyExists, _ := n.internalNetworkPolicyStore.Get(internalNetworkPolicyName)
+	if oldInternalPolicyExists {
+		oldInternalNetworkPolicy = oldInternalNetworkPolicyObj.(*antreatypes.NetworkPolicy)
+	}
+
+	// appliedToGroupsToSync tracks new AppliedToGroups created by this NetworkPolicy.
+	appliedToGroupsToSync := sets.NewString()
+
+	// Create the internal NetworkPolicy, AppliedToGroups and AddressGroups if they don't exist. They need to be updated
+	// atomically to avoid race conditions between workers that process multiple NetworkPolicies.
+	func() {
+		n.internalNetworkPolicyMutex.Lock()
+		defer n.internalNetworkPolicyMutex.Unlock()
+
+		if !oldInternalPolicyExists {
```

Comment:
Yes, they need to be protected by the same mutex because NP store has index of AddressGroups and AppliedToGroups, which is used to determine whether an AddressGroup/AppliedToGroup is orphan or not. For example, the following code means the AppliedToGroup is no longer used and can be deleted:
```
objs, _ := n.internalNetworkPolicyStore.GetByIndex(store.AppliedToGroupIndex, atgName)
if len(objs) == 0 {
``` 

We must ensure the AddressGroup/AppliedToGroup data (stored in their own store) and their index (stored in NP store) are updated atomically, otherwise the first race condition the PR description describes could happen:

1. addANP creates a group for ANP A;
2. addNetworkPolicy reuses the group for KNP B, is going to create an internal NetworkPolicy;
3. deleteANP deletes the group for ANP A because at that moment no other internal NetworkPolicies are using the group;
4. addNetworkPolicy commits the internal NetworkPolicy for KNP B to storage, but the group no longer exists.

Reply:
Ok. Got it.